egs/wsj/s5/steps/nnet3/make_bottleneck_features.sh

```diff
@@ -0,0 +1,123 @@
+nj=4
+cmd=queue.pl
+use_gpu=false
+ivector_dir=
+compress=true
+# End configuration options.
+
+echo "$0 $@"  # Print the command line for logging
+
+[ -f path.sh ] && . ./path.sh # source the path.
+. parse_options.sh || exit 1;
+
+if [ $# -lt 3 ]; then
+   echo "usage: steps/nnet3/dump_bottleneck_features.sh <bnf-node-name> <input-data-dir> <bnf-data-dir> <nnet-dir> [<log-dir> [<bnfdir>] ]"
+   echo "e.g.:  steps/nnet3/dump_bottleneck_features.sh tdnn_bn.renorm data/train data/train_bnf exp/nnet3/tdnn_bnf exp_bnf/dump_bnf bnf"
+   echo "Note: <log-dir> dafaults to <bnf-data-dir>/log and <bnfdir> defaults to"
+   echo " <bnf-data-dir>/data"
+   echo "main options (for others, see top of script file)"
+   echo "  --config <config-file>                           # config containing options"
+   echo "  --nj <nj>                                        # number of parallel jobs"
+   echo "  --cmd (utils/run.pl|utils/queue.pl <queue opts>) # how to run jobs."
+   echo "  --ivector-dir                                    # directory for ivectors"
+   exit 1;
+fi
+bnf_name=$1 # the component-node name in nnet3 model used for bottleneck feature extraction
+data=$2
+bnf_data=$3
+nnetdir=$4
+if [ $# -ge 4 ]; then
+  logdir=$4
+else
+  logdir=$bnf_data/log
+fi
+if [ $# -ge 5]; then
+  bnfdir=$4
+else
+  bnfdir=$bnf_data/data
+fi
+
+# Assume that final.nnet is in nnetdir
+cmvn_opts=`cat $nnetdir/cmvn_opts`;
+bnf_nnet=$nnetdir/final.raw
+if [ ! -f $bnf_nnet ] ; then
+  echo "$0: No such file $bnf_nnet";
+  exit 1;
+fi
+
+if $use_gpu; then
+  compute_queue_opt="--gpu 1"
+  compute_gpu_opt="--use-gpu=yes"
+  if ! cuda-compiled; then
+    echo "$0: WARNING: you are running with one thread but you have not compiled"
+    echo "   for CUDA.  You may be running a setup optimized for GPUs.  If you have"
+    echo "   GPUs and have nvcc installed, go to src/ and do ./configure; make"
+    exit 1
+  fi
+else
+  echo "$0: without using a GPU this will be very slow.  nnet3 does not yet support multiple threads."
+  compute_gpu_opt="--use-gpu=no"
+fi
+
+
+## Set up input features of nnet
+name=`basename $data`
+sdata=$data/split$nj
+
+mkdir -p $logdir
+mkdir -p $bnf_data
+echo $nj > $nnetdir/num_jobs
+splice_opts=`cat $nnetdir/splice_opts 2>/dev/null`
+[[ -d $sdata && $data/feats.scp -ot $sdata ]] || split_data.sh $data $nj || exit 1;
+
+if [ "$ivector_dir" != "" ];then
+  use_ivector=true
+fi
+
+feats="ark,s,cs:apply-cmvn $cmvn_opts --utt2spk=ark:$sdata/JOB/utt2spk scp:$sdata/JOB/cmvn.scp scp:$sdata/JOB/feats.scp ark:- |"
+ivector_feats="scp:utils/filter_scp.pl $sdata/JOB/utt2spk $ivector_dir/ivector_online.scp |"
+
+if [ $stage -le 1 ]; then
+  echo "$0: Generating bottleneck features using $bnf_nnet model as output of "
+  echo "    component-node with name $bnf_name."
+  echo "output-node name=output input=$bnf_name" > output.config
+  modified_bnf_nnet="nnet3-copy --edits='remove-output-nodes name=output' $bnf_nnet - | nnet3-copy --nnet-config=output.config - - |"
```

Comment:
I think output.config should go to $sdata (or somewhere else tied to this given running instance) but definitely should not stay in the current directory -- that creates opportunity for race conditions where, for example you will be running the script twice with different setting (for example, during experimentation with parameters)